syntax = "proto3";

package caffa.rpc;

option optimize_for = SPEED;
option java_multiple_files = true;
option java_package = "org.caffa.rpc";

service ObjectAccess
{
    rpc GetDocument( DocumentRequest ) returns( Object ) {}
    rpc ExecuteMethod( MethodRequest ) returns( Object ) {}
}

message Object
{
    string class_keyword = 1;
    uint64 address       = 2;
    string json          = 3;
}

message ObjectList
{
    repeated Object objects = 1;
}

message DocumentRequest
{
    string document_id = 1;
}

message DescendantsRequest
{
    Object self                      = 1;
    string descendants_class_keyword = 2;
}

message AncestorsRequest
{
    Object self                    = 1;
    string ancestors_class_keyword = 2;
}

message ChildRequest
{
    Object self       = 1;
    string field_name = 2;
}

message MethodRequest
{
    Object self   = 1;
    string method = 2;
    Object params = 3;
}
